File: Lab2/Hamming/Hamming2.java
package Hamming;

import java.util.ArrayList;
import java.util.Scanner;

import java.net.Socket;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.io.OutputStreamWriter;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.net.UnknownHostException;

class HammEmisor {

    private static String HOST = "127.0.0.1";
    private static int PORT = 5555;
    public String text = "";
    public String codedMessage = "";
    public String messageWithIntegrity = "";
    public String messageWithNoise = "";

    public HammEmisor(String text) {
        this.text = text;
    }

    public String requestText() {
        System.out.println("\nIngrese el texto a enviar: ");
        Scanner scanner = new Scanner(System.in);
        String text = scanner.nextLine();
        scanner.close();
        return text;
    }

    public String codeMessage(String text) {
        StringBuilder binary = new StringBuilder();
        
        for (int i = 0; i < text.length(); i++) {
            char character = text.charAt(i);
            int asciiValue = (int) character;
            
            String binaryString = Integer.toBinaryString(asciiValue);
            
            while (binaryString.length() < 8) {
                binaryString = "0" + binaryString;
            }
            
            binary.append(binaryString);
        }
        
        return binary.toString();
    }

    public int calculateParityBits(String text) {
        int m = text.length();
        int r = 0;

        while (Math.pow(2, r) < m + r + 1) {
            r++;
        }

        return r;
    }

    public String calculateIntegrity(String text) {
        int n = text.length() + calculateParityBits(text);
        HammingCoding hamming = new HammingCoding(n, text.length(), text);
        hamming.fullCoding();
        return hamming.hammingCode;
    }

    public double calculateNoiseProbability(String text) {
        int length = text.length();
        double errorsInInterval = length/2;

        double baseProb = 1 / errorsInInterval; //0.25
        return baseProb;

        // double randomValue = Math.random();
        // if (randomValue < 0.33) {
        //     return baseProb - adjustment;
        // } else if (randomValue < 0.66 && randomValue >= 0.33) {
        //     return baseProb + adjustment;
        // } else {
        //     return baseProb;
        // }
    }

    public String addNoise(String text) {
        double noiseProbability = calculateNoiseProbability(text);
        System.out.println("Probabilidad de ruido: " + noiseProbability);
        // noiseProbability = 0.08;
        StringBuilder noisyText = new StringBuilder();

        for (int i = 0; i < text.length(); i++) {
            char originalBit = text.charAt(i);
            char noisyBit;

            double randomValue = Math.random();

            //verificar si se voltean los bits
            if (randomValue < noiseProbability) {
                noisyBit = (originalBit == '0') ? '1' : '0'; 
            } else {
                noisyBit = originalBit; 
            }

            noisyText.append(noisyBit);
        }

        return noisyText.toString();
    }

    private void sendData(Socket socketClient, String originalData, String noisyData)
        throws IOException, InterruptedException {

        OutputStreamWriter writer = null;

        //crear payload
        String payload = originalData + "," + noisyData+ "\n";

        //mandar data
        // System.out.println("Enviando data");
        writer = new OutputStreamWriter(socketClient.getOutputStream());
        writer.write(payload);
        writer.flush(); // Flush the output stream to ensure data is sent
        Thread.sleep(100); // espera, opcional

        //limpieza
        // System.out.println("Liberando socket");
        // writer.close();
    }
    
    public void fullEmisor(Socket socketClient) {
        // this.text = this.requestText();
        this.codedMessage = this.codeMessage(this.text);
        // System.out.println("Mensaje codificado         : " + this.codedMessage);
        this.messageWithIntegrity = this.calculateIntegrity(this.codedMessage);
        this.messageWithNoise = this.addNoise(this.messageWithIntegrity);
        System.out.println("Mensaje con ruido          : " + this.messageWithNoise);

        try {
            sendData(socketClient, this.codedMessage, this.messageWithNoise);
        } catch (Exception e) {
            e.printStackTrace();
        }
}
    
}

class HammReceptor {

    // private final String HOST = "127.0.0.1";
    private final int PORT = 5555;
    private boolean errors = false;
    private String ogData = "";
    private String noisyData = "";
    private String originalData = "";
    private String decodedMessage = "";

    public String receive() throws IOException {
        String data = "";
        try (ServerSocket serverSocket = new ServerSocket(PORT)) {
            System.out.println("\nEsperando mensaje...");
            Socket socket = serverSocket.accept();
            System.out.println("Conexion aceptada de: " + socket.getInetAddress());

            byte[] buffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = socket.getInputStream().read(buffer)) != -1) {
                data += new String(buffer, 0, bytesRead);
            }
        }
        return data;
    }

    public int calculateParityBits(String data) {
        int m = data.length();
        int r = 0;
        while (Math.pow(2, r) < m + r + 1) {
            r++;
        }
        return r;
    }

    public void compareData(String ogData, String noisyData) {
        if (ogData.equals(noisyData)) {
            errors = false;
            System.out.println("\nIntegridad verficada y trama correcta!");
        } else {
            errors = true;
            System.out.println("\nIntegridad Incorrecta!");
        }
    }

    public String eliminateParityBits(String data) {
        StringBuilder originalData = new StringBuilder();
        for (int i = 1; i <= data.length(); i++) {
            if ((i & (i - 1)) != 0) {
                originalData.append(data.charAt(i - 1));
            }
        }
        return originalData.toString();
    }

    public String deCodeMessage(String data) {
        if (errors) {
            return "\nError: No se puede decodificar el mensaje!";
        } else {
            StringBuilder originalText = new StringBuilder();
            String[] binaryChunks = data.split("(?<=\\G.{8})");
            for (String chunk : binaryChunks) {
                int asciiValue = Integer.parseInt(chunk, 2);
                originalText.append((char) asciiValue);
            }
            return originalText.toString();
        }
    }

    public void fullReceptor() {
        try {
            String dataReceived = receive();
            String[] dataReceivedSplit = dataReceived.split(",");
            ogData = dataReceivedSplit[0];
            noisyData = dataReceivedSplit[1];
            System.out.println("\nTrama original            -> " + ogData);
            System.out.println("Trama con errores         -> " + noisyData);

            int n = calculateParityBits(ogData) + ogData.length();
            HammingCoding hamming = new HammingCoding(n, ogData.length(), ogData);
            hamming.fullCoding();
            
            try {
                HamingDecoding hammingDecoding = new HamingDecoding(hamming.hammingCode, noisyData);
                hammingDecoding.fullDecoding(hamming);
                compareData(hamming.hammingCode, hammingDecoding.correctData);
                originalData = eliminateParityBits(hammingDecoding.correctData);
                if (!errors) {
                    System.out.println("Trama corregida           -> " + originalData);
                    decodedMessage = deCodeMessage(originalData);
                    System.out.println("Mensaje decodificado      -> " + decodedMessage);
                } else {
                    System.out.println(deCodeMessage(noisyData));
                }
            } catch (Exception e) {
                errors = true;
                decodedMessage = deCodeMessage(noisyData);
                System.out.println(decodedMessage);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

class Hamming2 {

    public static void main(String[] args) {
        String HOST = "127.0.0.1";
        int PORT = 5555;
        // Scanner scanner = new Scanner(System.in);
        // System.out.println("\n----HAMMING----");
        // System.out.println("Que desea hacer?");
        // System.out.println("1. Emisor");
        // System.out.println("2. Receptor");
        // System.out.print("Opcion -> ");
        // int option = scanner.nextInt();
        int option = 1;

        if (option == 1) {
            ArrayList<String> text = new ArrayList<>();
            try (BufferedReader reader = new BufferedReader(new FileReader("tests.txt"))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    text.add(line);
                }
            } catch (IOException e) {
                e.printStackTrace();
                return; // Exit the program if there's an error reading the file
            }

            try {
                // Establish socket connection outside the loop
                Socket socketClient = new Socket(InetAddress.getByName(HOST), PORT);

                for (String t : text) {
                    HammEmisor sender = new HammEmisor(t);
                    sender.fullEmisor(socketClient);
                }

                socketClient.close(); // Close the socket connection after sending all messages
            } catch (IOException e) {
                e.printStackTrace();
            }
            System.out.println("Mensajes enviados.");
        } else if (option == 2) {
            
            HammReceptor receiver = new HammReceptor();
            receiver.fullReceptor();
            
        } else {
            System.out.println("Opcion Invalida.");
        }

        // scanner.close();
    }
}

File: Lab2/Hamming/Hamming2.py
import socket
import random
import math
from Hamming import *

class HammEmisor:

    def __init__(self):
        self.HOST = "127.0.0.1"
        self.PORT = 5555
        self.text = ""
        self.coded_message = ""
        self.message_with_integrity = ""
        self.message_with_noise = ""

    def request_text(self):
        text = input("\nEnter the text to send: ")
        return text

    def code_message(self, text):
        binary = ""

        for char in text:
            ascii_value = ord(char)
            binary_string = bin(ascii_value)[2:].zfill(8)
            binary += binary_string

        return binary

    def calculate_parity_bits(self, text):
        m = len(text)
        r = 0

        while 2 ** r < m + r + 1:
            r += 1

        return r

    def calculate_integrity(self, text):
        n = len(text) + self.calculate_parity_bits(text)
        hamming = HammingCoding(n, len(text), text)
        hamming.fullCoding()
        return hamming.hammingCode

    def calculate_noise_probability(self, text):
        length = len(text)
        errors_in_interval = length / 2

        base_prob = 1 / errors_in_interval
        adjustment = base_prob / 2

        random_value = random.random()
        if random_value < 0.5:
            return base_prob - adjustment
        else:
            return base_prob + adjustment

    def add_noise(self, text):
        noise_probability = self.calculate_noise_probability(text)
        noisy_text = ""

        for bit in text:
            random_value = random.random()

            if random_value < noise_probability:
                noisy_bit = '1' if bit == '0' else '0'
            else:
                noisy_bit = bit

            noisy_text += noisy_bit

        return noisy_text

    def send_data(self, original_data, noisy_data):
        print("\nStarting sender")
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as socket_client:
                socket_client.connect((self.HOST, self.PORT))
                payload = f"{original_data},{noisy_data}"
                print("Sending data")
                socket_client.sendall(payload.encode())
                socket_client.shutdown(socket.SHUT_WR)
                print("Socket released")
        except Exception as e:
            print(e)

    def full_emisor(self):
        self.text = self.request_text()
        self.coded_message = self.code_message(self.text)
        print("Coded message              :", self.coded_message)
        self.message_with_integrity = self.calculate_integrity(self.coded_message)
        self.message_with_noise = self.add_noise(self.message_with_integrity)
        print("Message with noise         :", self.message_with_noise)

        try:
            self.send_data(self.coded_message, self.message_with_noise)
        except Exception as e:
            print(e)

class HammReceptor():
    def __init__(self):
        self.HOST = "127.0.0.1"
        self.PORT = 5555
        self.errors = False
        self.ogData = ""
        self.noisyData = ""
        self.originalData = ""
        self.decodedMessage = ""

    def receive(self):
        data = ""
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            #asignar socket a ip: puerto especifico
            s.bind((self.HOST, self.PORT))
            s.listen()

            # esperar conexion
            conn, addr = s.accept()
            with conn:
                print(f"\nConexion entrante del proceso {addr}")
                while True:
                    data_received = conn.recv(1024)
                    if not data_received:
                        break # se termino de recibir todo
                    data += data_received.decode("utf-8")
                    print(f"\nRecibido: \n{data!r}")
        
        return data
    
    def calculateParityBits(self, data):
        m = len(data)
        r = 0
        while 2 ** r < m + r + 1:
            r += 1
        
        return r

    def compareData(self, ogData, noisyData):
        if ogData == noisyData:
            self.errors = False
            print("\nIntegridad verficada y trama correcta!")
        else:
            self.errors = True
            print("\nIntegridad Incorrecta!")
    
    def eliminateParityBits(self, data):
        original_data = ""

        # Iterate through the data and skip bits at positions that are powers of 2
        for i, bit in enumerate(data, start=1):
            if i & (i - 1) == 0:
                continue  # Skip bits at powers of 2
            original_data += bit

        return original_data

    def deCodeMessage(self, data):
        if self.errors:
            return "\nError: No se puede decodificar el mensaje!"
        else:
            # Split the data into 8-bit chunks
            binary_chunks = [data[i:i+8] for i in range(0, len(data), 8)]

            # Convert each binary chunk to ASCII value
            ascii_values = [int(chunk, 2) for chunk in binary_chunks]

            # Convert ASCII values to characters and join to form the original text
            original_text = ''.join(chr(value) for value in ascii_values)
            
            return original_text

    def fullReceptor(self):
        print("\nEsperando mensaje...")
        dataRecieved = self.receive()
        dataRecieved = dataRecieved.split(",")
        self.ogData = dataRecieved[0]
        self.noisyData = dataRecieved[1]
        print("\nTrama original            ->", self.ogData)
        print("Trama con errores         ->", self.noisyData)

        #codificando y decodificando con los algoritmos originales
        n = self.calculateParityBits(self.ogData) + len(self.ogData)
        hamming = HammingCoding(n, len(self.ogData), self.ogData)
        hamming.fullCoding()
        try:
            hammingDecoding = HammingDecoding(hamming.hammingCode, self.noisyData)
            hammingDecoding.fullDecoding(hamming)
            self.compareData(hamming.hammingCode, hammingDecoding.correctData)
            self.originalData = self.eliminateParityBits(hammingDecoding.correctData)
            if self.errors == False:
                print("Trama corregida      ->", self.originalData)
                self.decodedMessage = self.deCodeMessage(self.originalData)
                print("Mensaje decodificado ->", self.decodedMessage)
            else:
                print(self.decodedMessage(self.noisyData))
        except Exception as e:
            self.errors = True
            self.decodedMessage = self.deCodeMessage(self.noisyData)
            print(self.decodedMessage)

class Hamming2():
    def __init__(self):
        print("\nQue desea hacer?")
        print("1. Emisor")
        print("2. Receptor")

        opcion = int(input("Opcion -> "))

        if opcion == 1:
            self.emisor()
        elif opcion == 2:
            self.receptor()

    def emisor(self):
        emisor = HammEmisor()
        emisor.full_emisor()

    def receptor(self):
        receptor = HammReceptor()
        receptor.fullReceptor()

if __name__ == "__main__":
    Hamming2()